[FIRRTL] Add Dedup pass

youngar opened the pull request:
There is still a problem with memory port annotations and NLAs that I need to think through.

@seldridge  This is rebased off of `llvm/main` now that all prerequisites of this PR have been merged. This branch has a bug-fix that was not present earlier today.  It would be a good idea to work off this branch if anyone is working on porting the SFC dedup tests.

youngar commented:
Opened an issue regarding the memory port annotations: https://github.com/llvm/circt/issues/2526
My vote is to merge this before we fix it, since is should be impossible for this issue to be triggered right now. I think the representation of non-local port annotations is still needs a bit more work in general.

seldridge commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,985 @@
+    }
+  }
+
+  /// Private NLA creation method.  This function requires an attribute array
+  /// with a placeholder in the first element, where the root refence of the NLA
+  /// will be inserted. This clones the NLA for each instantiation of the
+  /// "fromModule".
+  SmallVector<FlatSymbolRefAttr>
+  createNLAsImpl(OpBuilder &builder, StringAttr toModuleName, AnnoTarget to,
+                 Operation *fromModule, SmallVectorImpl<Attribute> &namepath) {
```
Nit: should this create the NLAs at the top of the circuit as opposed to wherever the `builder` happens to be?

youngar replied:
I was trying to place them above the module that stays when two modules dedup, but it ends up looking random because i wrote the test modules in leaf-first order.  If i write the parent module first it looks a little cleaner.  I agree that this does seem a little random right now, do we have a preference for putting all the NLAs at the top of the circuit?   If so I guess I can put each test in its own circuit, otherwise I can swap the order of modules so it looks better.

seldridge replied:
It seems cleaner to aggregate them at the top of the circuit or at least before the first module that uses them. This is not a hard requirement, though.

(This also came from me writing tests... I had expected to the see the NLA op show up before any symbol uses, but it wasn't the case.)

Not a big deal.

youngar replied:
Fixed this up to put the NLAs at the top of the circuit. I ran into a problem forcing me to insert each NLA right at the beginning of the circuit, effectively leaving them in the reverse order they were created in.

seldridge commented:
SFC dedup tests are ported.  When you rebase, this PR can remove the `XFAIL` on [test/Dialect/FIRRTL/SFCTests/dedup.fir](https://github.com/llvm/circt/blob/main/test/Dialect/FIRRTL/SFCTests/dedup.fir).  The coverage provided by these tests is probably not great as they are primarily concerned with catching tricky situations involving SFC's [`firrtl.RenameMap`](https://www.chisel-lang.org/api/firrtl/1.5.0/firrtl/RenameMap.html). That said, these should bang on the NLA part of the pass pretty heavily (e.g., there's a 12-annotation test 😅).

The second test, [test/Dialect/FIRRTL/SFCTests/dedup-errors.fir](https://github.com/llvm/circt/blob/main/test/Dialect/FIRRTL/SFCTests/dedup-errors.fir) which will still fail with the current state of this PR is due to several reasons:
1. Instance port annotations need to be canonicalized in `LowerAnnotations` to live on the module.
2. Connect/Partial Connect expansion needs to respect flow.
3. The conversion of a `DontTouchAnnotation` to a symbol destroys some information that can't be perfectly recovered (see: #2531) and means `DontTouchAnnotation` is "sticky" in CIRCT where it is promoted to an NLA in SFC.

youngar commented:
This PR has another dependency: https://github.com/llvm/circt/pull/2550

youngar commented:
I fixed up the SFC dedup tests and removed the `XFAIL`.  I moved 1 test from `dedup-errors.fir` over to `dedup.fir`, specifically the one about not deduping when the port names differ and partial connect is used on an instance.

darthscsi commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+        nlaMap[element.cast<InnerRefAttr>().getModule()].push_back(nla);
+    }
+  }
+  return nlaMap;
```
Isn't this precisely the same information stored on the modules in the class=circt.nonlocal?

youngar replied:
Since the `circt.nonlocal` is stored on instances (not the module) its more equivalent to:
```c++
SmallVector getParticipatingNLAs(Module mod) {
  for (instance : instanceGraph[mod].uses) {
    for (anno : AnnotationSet(instance->getInstance()) {
      if (anno.class == "circt.nonlocal") {
        results.push_back(symbolTable[anno.getMember("circt.nonlocal"))
      }
    }
    return results;
  }
```

darthscsi commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+    sha.init();
+  }
+
+  void update(const void *pointer) {
```
This seems dangerous, what runtime-data is it capturing?

darthscsi replied:
Or rather, is there a narrow type that would be safer?

youngar replied:
This is for anytime I call `getOpaquePointer()`, which returns a `void *`.  RIght now this is capturing: `TypeID`, `Type`, `Attribute`, `OperationName`.  I could duplicate this function for each one to make it safer, would that be better?

darthscsi commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+    sha.update(ArrayRef(addr, sizeof pointer));
+  }
+
+  void update(unsigned value) {
```
Maybe use size_t or unsigned long long to capture all integers?

darthscsi commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+  /// replaces them with new partial connects.
+  using LazyValue = llvm::function_ref<Value(ImplicitLocOpBuilder &)>;
+  // NOLINTNEXTLINE(misc-no-recursion)
+  void fixupPartialConnect(ImplicitLocOpBuilder &builder, LazyValue dst,
```
Shouldn't the bundle types be compatible for the hash to match, in that each field type matches?   The you know field 0 -> field 0, field 1 -> field 1, etc.  You shouldn't need to search by name.  It seems, you should be able to do a bitcast and be done with it.

darthscsi commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+
+  /// When we replace a bundle type with a similar bundle with different field
+  /// names, we have to rewrite all the code to use the new field names. This
+  /// mostly affects subfield result types and any bulk connects.
```
Bitcast and a canonicalization rule?  (sub "foo" (bitcast srcStruct)) -> sub ("bar" srcStruct).

darthscsi commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+
+  void update(TypeID typeID) { update(typeID.getAsOpaquePointer()); }
+
+  void update(BundleType type) {
```
This seems like it should preclude lots of the connect issues?

darthscsi reviewed (approved):
Let's do this!

seldridge commented on test/Dialect/FIRRTL/dedup.mlir:
```diff
@@ -0,0 +1,411 @@
+  }
+}
+
+// CHECK-LABEL: firrtl.circuit "Annotations"
+firrtl.circuit "Annotations" {
```
This test hits on an angle that never occurred to me... When merging two modules that have multiple possible match points, it does bring up a question about if annotations should factor in to help disambiguate points which can match multiple ways. I.e., any of the wires could theoretically be deduplicated to any other wire and it is possible that a "lower energy" representation could be achieved by using common annotations as hints for deduplication. Don't do this as it complicates the problem substantially and I expect any benefit is balanced by the fact that this is an exceedingly rare occurrence.

seldridge commented on test/Dialect/FIRRTL/dedup.mlir:
```diff
@@ -0,0 +1,411 @@
+}
+
+
+// Don't deduplicate modules with NoDedup.
+// CHECK-LABEL: firrtl.circuit "NoDedup"
+firrtl.circuit "NoDedup" {
+  firrtl.module @Simple() { }
+  // CHECK: firrtl.module @NoDedup 
+  firrtl.module @NoDedup() attributes {annotations = [{class = "firrtl.transforms.NoDedupAnnotation"}]} { }
+}
```
Can this test be modified to place the `NoDedupAnnotation` or a submodule which would dedup with another submodule?

This test inadvertently hits a corner case where I don't think the main module is supposed to dedup with anything else. This is getting into the realm of pedantically-exact SFC behavior, but the SFC does do this. Consider:

```scala
circuit Foo:
  module Baz:
    input a: UInt<1>
    output b: UInt<1>

    b <= a

  module Bar:
    input a: UInt<1>
    output b: UInt<1>

    b <= a

  module Foo:
    input a: UInt<1>
    output b: UInt<1>

    b <= a
```

This will result in `Baz` and `Bar` deduping, but not with `Foo` using `firrtl -i Foo.fir -cll firrtl.transforms.DedupModules:trace`:

```scala
circuit Foo :
  module Baz :
    input a : UInt<1>
    output b : UInt<1>

    b <= a

  module Foo :
    input a : UInt<1>
    output b : UInt<1>

    b <= a
```

I would add a test that the main module can't dedup with anything, as well.

youngar replied:
So it turns out that this test is completely broken.

One thing we talked about in the past is that it should be *illegal* for any module to be unreachable from the top of the hierarchy.  This means that it should actually be impossible for the "main module" to dedup with anything, unless we had a recursive module instantiation (also illegal).  I'm not sure how hard we are on this stance, and we definitely never implemented any verification for this.

With that in mind, the dedup pass never actually sees modules that are unreachable.  This is because we visit modules leaf-first using `llvm::post_order(mainModule)`, which implicitly does not include unreachable modules.

So I have updated this test as you suggested above, a single top module instantiating two submodule.  I can add the second test suggested, and it would work because unreachable modules are never deduped with anything, or I can omit it because it's illegal.

seldridge replied:
It's fine to just update the existing test. I just wanted to make sure that we weren't opening the door for deduplication with out-of-tree modules (which the test seemed to be relying on to work). 😉

seldridge reviewed (approved):
This is a monstrous piece of work. Awesome job.

Great tests. Those are extremely comprehensive.

I have some minor comments throughout and am fine for this to land anytime.

youngar commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+using namespace firrtl;
+using hw::InnerRefAttr;
+
+/// This stores a mapping from a module name to every NLA that it particiapates
+/// in.
+using NLAMap = DenseMap<Attribute, std::vector<NonLocalAnchor>>;
```
When we go through `addAnnotationContext` we will be cloning each NLA in the vector for every parent module.  When we create the new NLA, it needs to be recorded in the `NLAMap`, which can cause it to grow. Since the `NLAMap` grows while we're iterating the NLA vector stored in the map, we will crash if we use a `SmallVector` but we're safe with a `std::vector.` This is honestly a bit sketchy, there is some disagreement if moving a `std::vector` does not invalidate iterators, but this is done in a couple places over LLVM and a lot of things would break otherwise.

youngar commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -0,0 +1,1009 @@
+                         FModuleOp toModule, FModuleOp fromModule) {
+    addAnnotationContext(renameMap, toModule, toModule);
+    addAnnotationContext(renameMap, toModule, fromModule);
+  }
```
`addAnnotationContext` actually does two things,
1. renaming the module in an NLA if `fromModule != toModule`, and 
2. adds more context to any NonLocalAnchor rooted at the `fromModule`.

We need to do `2` for both modules when they are deduped with each other.
